### ISSUE TITLE
Read command-line options from spec file

### ISSUE THREAD

**Issue by sveinse** [pyinstaller/pyinstaller#4482]
I have an application that I'm building twice: a onefile version and a onedir version. The spec files for doing this are very similar, but there are some differences. E.g. onedir calls EXE and COLLECT, while onefile calls EXE with some other options.

Is there a way I can programmically detect the `--onefile` or `--onedir` command-line option in the spec-file code? This way I would only need to maintain one spec file.

**Comment by danyeaw** [comment]
@sveinse I don't think this is currently a supported feature.

**Comment by htgoebel** [comment]
Interesting point. You should at least be able to access `sys.argv` from within the .spec-file.

If you think, this might be useful for other, to, I suggest to implement a more generic solution and provide a pull-request.
You could add an element `ARGUMENTS` (name to be discussed) to the namespace of the spec-file (like there are currently e.g. `PYZ`, `EXE`) containing all (or the relevant= command line arguments. For course this need to be documented in the manual, too.

The place to implement this is around https://github.com/pyinstaller/pyinstaller/blob/develop/PyInstaller/building/build_main.py#L642. You would add the new element to `spec_namespace`.
Most command line-args result into a value in `CONF`. 

I'm lloking forwar for the pull-request :-). See our Development Guide on [how to create pull-requests for PyInstaller](https://pyinstaller.readthedocs.io/en/stable/development/pull-requests.html). Thanks.

**Comment by DemiBSel** [comment]
I am not sure how this would/should play out with the ArgumentParser rejecting unknown options and now that #6660 prevents non-default option values to be used along with .spec files.
Allowing command-line options would still be a convenient feature to avoid having to choose between multiple .spec files.

**Comment by bwoodsend** [comment]
What would you do with these extra command line arguments? I can't say I'm excited by the idea that users would then be reporting failing build commands with random user defined arguments involved.

**Comment by DemiBSel** [comment]
As @sveinse said, this would allow using only one .spec file to bundle one app in multiple ways, enabling the console without modifying the spec file.
Might be made safer if only arguments with a specific prefix are allowed ?

**Comment by bwoodsend** [comment]
There's always the option to squirt parametrisation in via environment variables.

> Might be made safer if only arguments with a specific prefix are allowed ?

That would be less messy. I wouldn't mind that being a feature.

I have this vague idea which I doubt I will ever implement that the spec file should just be a self contained Python script with `from PyInstaller import Analysis, EXE, PYZ, Bundle` at the top which you'd then invoke with `python your-code.spec`. It would require purging the hidden `CONF` class but it would solve this issue because you'd be free to stick your own CLI parser at the top and do whatever you want with the parameters.

**Comment by bwoodsend** [comment]
What actually is the required scope for this? i.e. How complex is the information you want to get into the spec file? If all you need is some simple string or boolean switches (e.g. `--debug-mode=on` or `--include-some-optional-dependency=true`) then this is a lot more achievable  than arbitrary argparse-like options.

[gradle](https://docs.gradle.org/) (a build system) has a `-P` flag which you can use to pass in string key/value pairs e.g. `-Pdebug_mode=1` which makes the variable available in its build scripts. We could have something similar provided that people don't need to pass anything more complex in than that?

**Comment by rokm** [comment]
What is being asked for in the original request can be readily achieved using environment variables, though: [example](https://github.com/pyinstaller/pyinstaller/blob/develop/tests/functional/specs/pyi_osx_aevent_handling_carbon.spec). (And as a sidenote, passing `--onefile` or `--onedir` on CLI when building with spec now raises an error, so trying to infer build mode from CLI arguments in the spec file is rather moot).

I suppose one could pass whole custom/additional command-line string via an environment variable, split it in the spec file using `shlex.split()` and feed it to `argparse`, without having to touch PyInstaller's CLI option parsing...

**Comment by bwoodsend** [comment]
> I suppose one could pass whole custom/additional command-line string via an environment variable, split it in the spec file using shlex.split() and feed it to argparse, without having to touch PyInstaller's CLI option parsing...

Well then you've got the joys nested quoting/escaping to navigate. I'd rather use something that's at least semi-platform/shell independent by default.

**Comment by rokm** [comment]
> > I suppose one could pass whole custom/additional command-line string via an environment variable, split it in the spec file using shlex.split() and feed it to argparse, without having to touch PyInstaller's CLI option parsing...
> 
> Well then you've got the joys nested quoting/escaping to navigate. I'd rather use something that's at least semi-platform/shell independent by default.

That's true as well. A switch for passing string key/value pairs (and making those available within spec via a global dict) sounds like a reasonable approach, then. The user can convert strings to their actual types (or even transform the whole dictionary to a list that can be fed to argparse).

**Comment by JamesPeters86** [comment]
I'm interested in working on this, but I'm having some difficulty with the test suites.

**Comment by arinwt** [comment]
My understanding of the underlying goal of the original request is to prevent duplicating spec info/logic. Another way to achieve this would be on spec inheriting from another spec. That's how this type of thing is achieved in other languages, and it might make more sense with the typical usage of calling pyinstaller with only the spec file.

**Comment by KarelChanivecky** [comment]
How about ignoring any arguments past special argument "--". The remaining args are ignored by pyinstaller, or assigned to a variable available during .spec runtime. Or the user can just search for this from argv and parse it however he likes. The "--"  arg is usually associated with "stop parsing" behaviour.

**Comment by bwoodsend** [comment]
Oh that's a good idea. Then the person writing the spec file is free to use argparse (or even click) which gives arbitrary types and sensible error handling if the user supplies an option that doesn't exist. It would even mean that `pyinstaller foo.spec -- --help` would show the available options so you don't have to dig through source code to find them.
